step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractRunPackagedAutomationPackagesMojo.java

```diff
@@ -29,41 +30,89 @@
+		// 2. Resolve additional library if required
+		String libResourceId = resolveAndUploadLibraryResource();
+		if (libResourceId != null) {
+			// add lib resource id
+			executionContext.put("libResourceId", libResourceId);
+		}
+
+		// 3. Execute just uploaded artifact in Step
 		executeBundleOnStep(executionContext);
 	}
 
-	protected String uploadResourceToStep() throws MojoExecutionException {
+	protected String resolveAndUploadLibraryResource() throws MojoExecutionException {
+		// for packaged automation packages we support 2 ways to define library file: by step resource id and as maven artifact
+		String libResourceId = null;
+		Map<String, String> libStepResourceSearchCriteria = getLibStepResourceSearchCriteria();
+		if (libStepResourceSearchCriteria != null && !libStepResourceSearchCriteria.isEmpty()) {
+			// reference library via resource id explicitly
+			libResourceId = resolveKeywordLibResourceByCriteria(libStepResourceSearchCriteria);
+		} else if (getLibArtifactId() != null && !getLibArtifactId().isEmpty()) {
+			getLog().info("Using maven artifact " + getLibArtifactGroupId() + ":" + getLibArtifactId() + ":" + getLibArtifactVersion() + " as library file");
+
+			// use deployed artifact as resource
+			org.eclipse.aether.artifact.Artifact remoteLibArtifact = getRemoteArtifact(getLibArtifactGroupId(), getLibArtifactId(), getLibArtifactVersion(), getLibArtifactClassifier(), "jar");
+			if (remoteLibArtifact == null) {
+				throw new MojoExecutionException("Library artifact is not resolved");
+			}
+
+			try {
+				// checkForDuplicates = true to restrict uploading the same library twice
+				libResourceId = uploadResourceToStep(remoteLibArtifact.getFile(), true);
+			} catch (SimilarResourceExistingException e) {
+				if (e.getSimilarResources() != null && !e.getSimilarResources().isEmpty()) {
+					getLog().info("Existing similar library resource is detected and will be reused as library file");
+					libResourceId = e.getSimilarResources().get(0).getId().toString();
+				}
+			}
+		}
+		if (libResourceId != null) {
+			getLog().info("The following Step resource will be used as keyword library: " + libResourceId);
+		}
+		return libResourceId;
+	}
+
+	protected String uploadResourceToStep(File fileToUpload, boolean checkForDuplicates) throws MojoExecutionException, SimilarResourceExistingException {
 		try (RemoteResourceManager resourceManager = createRemoteResourceManager()) {
-			File fileToUpload = getFileToUpload();
 			if(fileToUpload == null){
 				throw logAndThrow("Unable to detect an artifact to upload", getDefaultMojoException());
 			} else {
 				getLog().info("Artifact is detected for upload to Step: " + fileToUpload.getName());
 
-				Resource uploaded = resourceManager.createResource("temp", new FileInputStream(fileToUpload), fileToUpload.getName(), false, null);
+				Resource uploaded = resourceManager.createResource("temp", new FileInputStream(fileToUpload), fileToUpload.getName(), checkForDuplicates, null);
```

Comment:
@iegorov777 For keyword jars and libraries the resource type "functions" should be used instead of "temp". "temp" is reserved for temporary resources

Reply:
@jeromecomte 
OK. I've changed it to "functions".